File: sem4/MPP/mpp-proiect-repository-mariaruncan/ProjJavaFX/Persistence/src/main/java/app/persistence/repository/jdbc/TicketPurchasesDBRepository.java
package app.persistence.repository.jdbc;

import app.model.Game;
import app.model.Team;
import app.model.TicketPurchase;
import app.persistence.TicketPurchasesRepository;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.springframework.stereotype.Component;

import java.io.IOException;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

@Component
public class TicketPurchasesDBRepository implements TicketPurchasesRepository {

    private final JdbcUtils dbUtils;
    private static final Logger logger = LogManager.getLogger();

    public TicketPurchasesDBRepository(Properties props){
        logger.info("Initializing TicketPurchasesDBRepository with properties: {} ", props);
        dbUtils = new JdbcUtils(props);
    }

    public TicketPurchasesDBRepository(){
        Properties props = new Properties();
        try {
            props.load(EmployeesDBRepository.class.getResourceAsStream("/bd.properties"));
            System.out.println("Server properties set. ");
            props.list(System.out);
        } catch (IOException e) {
            System.err.println("Cannot find server.properties " + e);
            dbUtils = null;
            return;
        }
        dbUtils = new JdbcUtils(props);
    }

    @Override
    public void add(TicketPurchase elem) {
        logger.traceEntry("saving task {} ", elem);
        Connection con = dbUtils.getConnection();

        try(PreparedStatement preStm = con.prepareStatement("INSERT INTO TicketPurchases (gameId, noOfTickets, " +
                "totalPrice, clientName) VALUES (?, ?, ?, ?);")){
            preStm.setInt(1, Math.toIntExact(elem.getGame().getId()));
            preStm.setInt(2, elem.getNoOfTickets());
            preStm.setDouble(3, elem.getTotalPrice());
            preStm.setString(4, elem.getClientName());
            int result = preStm.executeUpdate();
            logger.trace("Saved {} instances", result);
        }
        catch(SQLException ex){
            logger.error(ex);
            System.err.println("Error DB" + ex);
        }
        logger.traceExit();
    }

    @Override
    public void delete(Long id) {
        logger.traceEntry("deleting task {} ", id);
        Connection con = dbUtils.getConnection();

        try(PreparedStatement preStm = con.prepareStatement("DELETE FROM TicketPurchases WHERE id = ?;")){
            preStm.setInt(1, Math.toIntExact(id));
            int result = preStm.executeUpdate();
            logger.trace("Deleted {} instances", result);
        }
        catch(SQLException ex){
            logger.error(ex);
            System.err.println("Error DB" + ex);
        }
        logger.traceExit();
    }

    @Override
    public void update(Long aLong, TicketPurchase elem) {
        logger.traceEntry("updating task {} ", elem);
        Connection con = dbUtils.getConnection();

        try(PreparedStatement preStm = con.prepareStatement("UPDATE TicketPurchases SET gameId = ?, noOfTickets = ?," +
                "totalPrice = ?, clientName = ? WHERE id = ?;")){
            preStm.setInt(1, Math.toIntExact(elem.getGame().getId()));
            preStm.setInt(2, elem.getNoOfTickets());
            preStm.setDouble(3, elem.getTotalPrice());
            preStm.setString(4, elem.getClientName());
            preStm.setInt(5, Math.toIntExact(aLong));
            int result = preStm.executeUpdate();
            logger.trace("Updated {} instances", result);
        }
        catch(SQLException ex){
            logger.error(ex);
            System.err.println("Error DB" + ex);
        }
        logger.traceExit();
    }

    @Override
    public TicketPurchase findById(Long aLong) {
        logger.traceEntry();
        Connection con = dbUtils.getConnection();
        TicketPurchase ticketPurchase = null;

        try(PreparedStatement preStm = con.prepareStatement("SELECT * FROM TicketPurchases TP\n" +
                "WHERE id = ? LIMIT 1")){
            preStm.setInt(1, Math.toIntExact(aLong));
            try(ResultSet result = preStm.executeQuery()){
                if(result.next()){
                    Long tpId = result.getLong("TP.id");
                    long gameId = result.getLong("gameId");
                    Game game;
                    try(PreparedStatement preStm1 = con.prepareStatement("""
                            SELECT * FROM Games G
                            INNER JOIN Teams T1 on G.team1Id = T1.id
                            INNER JOIN Teams T2 on G.team2Id = T2.id
                            WHERE G.id = ?
                            LIMIT 1;""")){
                        preStm1.setInt(1, Math.toIntExact(gameId));
                        try(ResultSet result1 = preStm1.executeQuery()){

                            Long team1Id = result1.getLong("T1.id");
                            String team1Name = result1.getString("T1.name");
                            Team team1 = new Team(team1Name);
                            team1.setId(team1Id);

                            Long team2Id = result1.getLong("T2.id");
                            String team2Name = result1.getString("T2.name");
                            Team team2 = new Team(team2Name);
                            team2.setId(team2Id);

                            Long gameIdd = result1.getLong("G.id");
                            String description = result1.getString("description");
                            Integer totalNoOfTickets = result1.getInt("totalNoOfTickets");
                            Integer soldTickets = result1.getInt("soldTickets");
                            Float pricePerTicket = result1.getFloat("pricePerTicket");
                            LocalDate date =result1.getDate("date").toLocalDate();

                            game = new Game(team1, team2, date, description, totalNoOfTickets, soldTickets, pricePerTicket);
                            game.setId(gameIdd);
                        }
                    }
                    Integer noOfTickets = result.getInt("noOfTickets");
                    Float totalPrice = result.getFloat("totalPrice");
                    String clientName = result.getString("clientName");

                    ticketPurchase = new TicketPurchase(game, noOfTickets, totalPrice, clientName);
                    ticketPurchase.setId(tpId);
                }
            }
        }
        catch(SQLException ex){
            logger.error(ex);
            System.err.println("Error DB" + ex);
        }
        logger.traceExit();
        return ticketPurchase;
    }

    @Override
    public Iterable<TicketPurchase> findAll() {
        logger.traceEntry();
        Connection con = dbUtils.getConnection();
        List<TicketPurchase> ticketPurchases = new ArrayList<>();

        try(PreparedStatement preStm = con.prepareStatement("SELECT * FROM TicketPurchases TP;")){
            try(ResultSet result = preStm.executeQuery()){
                while(result.next()){
                    Long tpId = result.getLong(1);
                    long gameId = result.getLong("gameId");
                    Game game;
                    try(PreparedStatement preStm1 = con.prepareStatement("""
                            SELECT * FROM Games G
                            INNER JOIN Teams T1 on G.team1Id = T1.id
                            INNER JOIN Teams T2 on G.team2Id = T2.id
                            WHERE G.id = ?
                            LIMIT 1;""")){
                        preStm1.setInt(1, Math.toIntExact(gameId));
                        try(ResultSet result1 = preStm1.executeQuery()){

                            Long team1Id = result1.getLong("team1Id");
                            String team1Name = result1.getString(9);
                            Team team1 = new Team(team1Name);
                            team1.setId(team1Id);

                            Long team2Id = result1.getLong("team2Id");
                            String team2Name = result1.getString(11);
                            Team team2 = new Team(team2Name);
                            team2.setId(team2Id);

                            Long gameIdd = result1.getLong(1);
                            String description = result1.getString("description");
                            Integer totalNoOfTickets = result1.getInt("totalNoOfTickets");
                            Integer soldTickets = result1.getInt("soldTickets");
                            Float pricePerTicket = result1.getFloat("pricePerTicket");
                            LocalDate date = result1.getDate("date").toLocalDate();

                            game = new Game(team1, team2, date, description, totalNoOfTickets, soldTickets, pricePerTicket);
                            game.setId(gameIdd);
                        }
                    }
                    Integer noOfTickets = result.getInt("noOfTickets");
                    Float totalPrice = result.getFloat("totalPrice");
                    String clientName = result.getString("clientName");

                    TicketPurchase ticketPurchase = new TicketPurchase(game, noOfTickets, totalPrice, clientName);
                    ticketPurchase.setId(tpId);
                    ticketPurchases.add(ticketPurchase);
                }
            }
        }
        catch(SQLException ex){
            logger.error(ex);
            System.err.println("Error DB" + ex);
        }
        logger.traceExit();
        return ticketPurchases;
    }
}

File: sem4/MPP/mpp-proiect-repository-mariaruncan/ProjCSharp/Persistence/db/TicketPurchasesDBRepository.cs
using log4net;
using System;
using System.Collections.Generic;
using System.Data;
using System.Data.SQLite;
using Model;

namespace Persistence.db
{
    public class TicketPurchasesDBRepository : TicketPurchesesRepository
    {
        private static readonly ILog logger = LogManager.GetLogger(typeof(EmployeesDBRepository));
        DBUtils dbUtils;

        public TicketPurchasesDBRepository()
        {
            logger.Info("Initializing TicketPurchasesDBRepository");
            dbUtils = new DBUtils();
        }

        public void add(TicketPurchase elem)
        {
            logger.Info($"saving ticket purchase {elem}");
            IDbConnection con = dbUtils.GetConnection();

            try
            {
                using (IDbCommand cmd = con.CreateCommand())
                {
                    cmd.CommandText = "INSERT INTO TicketPurchases (gameId, noOfTickets, totalPrice, clientName) " +
                        "VALUES (@gameId, @noOfTickets, @totalPrice, @clientName);";

                    IDbDataParameter paramGameId = cmd.CreateParameter();
                    paramGameId.ParameterName = "@gameId";
                    paramGameId.Value = elem.Game.Id;
                    cmd.Parameters.Add(paramGameId);

                    IDbDataParameter paramNoOfTickets = cmd.CreateParameter();
                    paramNoOfTickets.ParameterName = "@noOfTickets";
                    paramNoOfTickets.Value = elem.NoOfTickets;
                    cmd.Parameters.Add(paramNoOfTickets);

                    IDbDataParameter paramTotalPrice = cmd.CreateParameter();
                    paramTotalPrice.ParameterName = "@totalPrice";
                    paramTotalPrice.Value = elem.TotalPrice;
                    cmd.Parameters.Add(paramTotalPrice);

                    IDbDataParameter paramClientName = cmd.CreateParameter();
                    paramClientName.ParameterName = "@clientName";
                    paramClientName.Value = elem.ClientName;
                    cmd.Parameters.Add(paramClientName);

                    var result = cmd.ExecuteNonQuery();
                    if (result == 0)
                        throw new RepositoryException("No ticket purchase added!");
                    logger.Info($"saved {result} instances");
                }
            }
            catch(SQLiteException ex)
            {
                logger.Error(ex.Message);
                Console.WriteLine("Error DB " + ex.Message);
            }
        }

        public void delete(long id)
        {
            logger.Info($"deleting ticket purchase {id}");
            IDbConnection con = dbUtils.GetConnection();

            try
            {
                using (IDbCommand cmd = con.CreateCommand())
                {
                    cmd.CommandText = "DELETE FROM TicketPurchases WHERE id = @id;";
                    IDbDataParameter paramId = cmd.CreateParameter();
                    paramId.ParameterName = "@id";
                    paramId.Value = id;
                    cmd.Parameters.Add(paramId);

                    var dataR = cmd.ExecuteNonQuery();
                    if (dataR == 0)
                        throw new RepositoryException("No ticket purchase deleted!");
                    logger.Info("exit delete EmployeeRepo");
                }
            }
            catch (SQLiteException ex)
            {
                logger.Error(ex.Message);
                Console.WriteLine("Error DB " + ex.Message);
            }
        }

        public IEnumerable<TicketPurchase> findAll()
        {
            logger.Info("enter findAll");
            IDbConnection con = dbUtils.GetConnection();
            IList<TicketPurchase> ticketPurchases = new List<TicketPurchase>();

            try
            {
                using (IDbCommand cmd = con.CreateCommand())
                {
                    cmd.CommandText = "SELECT TP.id, TP.gameId, TP.noOfTickets, TP.totalPrice, " +
                        "TP.clientName " +
                        "FROM TicketPurchases TP; ";

                    using (var result = cmd.ExecuteReader())
                    {
                        while (result.Read())
                        {
                            long tpId = result.GetInt64(0);
                            long gameId = result.GetInt64(1);
                            int noOfTickets = result.GetInt32(2);
                            float totalPrice = result.GetFloat(3);
                            String clientName = result.GetString(4);

                            Game game = null;

                            using (IDbCommand cmd1 = con.CreateCommand())
                            {
                                cmd1.CommandText = "SELECT G.id, G.description, G.totalNoOfTickets, G.soldTickets, " +
                                        "G.pricePerTicket, G.date, T1.id, T1.name, T2.id, T2.name " +
                                        "FROM Games G " +
                                        "INNER JOIN Teams T1 on G.team1Id = T1.id " +
                                        "INNER JOIN Teams T2 on G.team2Id = T2.id " +
                                        "WHERE G.id = @id " +
                                        "LIMIT 1;";
                                IDbDataParameter paramId = cmd1.CreateParameter();
                                paramId.ParameterName = "@id";
                                paramId.Value = gameId;
                                cmd1.Parameters.Add(paramId);

                                using (var result1 = cmd1.ExecuteReader())
                                {
                                    if (result1.Read())
                                    {
                                        long idG = result1.GetInt64(0);
                                        String description = result1.GetString(1);
                                        int totalNoOfTickets = result1.GetInt32(2);
                                        int soldTickets = result1.GetInt32(3);
                                        float pricePerTicket = result1.GetFloat(4);
                                        DateTime date = result1.GetDateTime(5);
                                        long team1Id = result1.GetInt64(6);
                                        String team1Name = result1.GetString(7);
                                        long team2Id = result1.GetInt64(8);
                                        String team2Name = result1.GetString(9);
                                        Team team1 = new Team(team1Name);
                                        team1.Id = team1Id;
                                        Team team2 = new Team(team2Name);
                                        team2.Id = team2Id;
                                        game = new Game(team1, team2, date, description, totalNoOfTickets, soldTickets, pricePerTicket);
                                        game.Id = gameId;
                                    }
                                }
                            }

                            TicketPurchase tp = new TicketPurchase(game, noOfTickets, totalPrice, clientName);
                            tp.Id = tpId;
                            ticketPurchases.Add(tp);
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                logger.Error(ex.Message);
                Console.WriteLine("Error DB " + ex.Message);
            }
            logger.Info("exit findAll");
            return ticketPurchases;
        }

        public TicketPurchase findById(long id)
        {
            logger.Info("enter findById");
            IDbConnection con = dbUtils.GetConnection();
            TicketPurchase tp = null;

            try
            {
                using (IDbCommand cmd = con.CreateCommand())
                {
                    cmd.CommandText = "SELECT TP.id, TP.gameId, TP.noOfTickets, TP.totalPrice, " +
                        "TP.clientName " +
                        "FROM TicketPurchases TP " +
                        "LIMIT 1;";

                    using (var result = cmd.ExecuteReader())
                    {
                        if (result.Read())
                        {
                            long tpId = result.GetInt64(0);
                            long gameId = result.GetInt64(1);
                            int noOfTickets = result.GetInt32(2);
                            float totalPrice = result.GetFloat(3);
                            String clientName = result.GetString(4);

                            Game game = null;

                            using (IDbCommand cmd1 = con.CreateCommand())
                            {
                                cmd1.CommandText = "SELECT G.id, G.description, G.totalNoOfTickets, G.soldTickets, " +
                                        "G.pricePerTicket, G.date, T1.id, T1.name, T2.id, T2.name " +
                                        "FROM Games G " +
                                        "INNER JOIN Teams T1 on G.team1Id = T1.id " +
                                        "INNER JOIN Teams T2 on G.team2Id = T2.id " +
                                        "WHERE G.id = @id;";
                                IDbDataParameter paramId = cmd1.CreateParameter();
                                paramId.ParameterName = "@id";
                                paramId.Value = gameId;
                                cmd1.Parameters.Add(paramId);

                                using (var result1 = cmd1.ExecuteReader())
                                {
                                    if (result1.Read())
                                    {
                                        long idG = result1.GetInt64(0);
                                        String description = result1.GetString(1);
                                        int totalNoOfTickets = result1.GetInt32(2);
                                        int soldTickets = result1.GetInt32(3);
                                        float pricePerTicket = result1.GetFloat(4);
                                        DateTime date = result1.GetDateTime(5);
                                        long team1Id = result1.GetInt64(6);
                                        String team1Name = result1.GetString(7);
                                        long team2Id = result1.GetInt64(8);
                                        String team2Name = result1.GetString(9);
                                        Team team1 = new Team(team1Name);
                                        team1.Id = team1Id;
                                        Team team2 = new Team(team2Name);
                                        team2.Id = team2Id;
                                        game = new Game(team1, team2, date, description, totalNoOfTickets, soldTickets, pricePerTicket);
                                        game.Id = gameId;
                                    }
                                }
                            }
                            tp = new TicketPurchase(game, noOfTickets, totalPrice, clientName);
                            tp.Id = tpId;
                        }
                    }
                }
            }
            catch(SQLiteException ex)
            {
                logger.Error(ex.Message);
                Console.WriteLine("Error DB " + ex.Message);
            }
            logger.InfoFormat("exit findById");
            return tp;
        }


        public void update(TicketPurchase elem, long id)
        {
            logger.Info($"updating ticket purchase {elem}");
            IDbConnection con = dbUtils.GetConnection();

            try
            {
                using (IDbCommand cmd = con.CreateCommand())
                {
                    cmd.CommandText = "UPDATE TicketPurchases SET gameId = @gameId, noOfTickets = @noOfTickets," +
                    "totalPrice = @totalPrice, clientName = @clientName WHERE id = @id;";

                    IDbDataParameter paramGameId = cmd.CreateParameter();
                    paramGameId.ParameterName = "@gameId";
                    paramGameId.Value = elem.Game.Id;
                    cmd.Parameters.Add(paramGameId);

                    IDbDataParameter paramNoOfTickets = cmd.CreateParameter();
                    paramNoOfTickets.ParameterName = "@noOfTickets";
                    paramNoOfTickets.Value = elem.NoOfTickets;
                    cmd.Parameters.Add(paramNoOfTickets);

                    IDbDataParameter paramTotalPrice = cmd.CreateParameter();
                    paramTotalPrice.ParameterName = "@totalPrice";
                    paramTotalPrice.Value = elem.TotalPrice;
                    cmd.Parameters.Add(paramTotalPrice);

                    IDbDataParameter paramClientName = cmd.CreateParameter();
                    paramClientName.ParameterName = "@clientName";
                    paramClientName.Value = elem.ClientName;
                    cmd.Parameters.Add(paramClientName);

                    IDbDataParameter paramId = cmd.CreateParameter();
                    paramId.ParameterName = "@id";
                    paramId.Value = elem.Id;
                    cmd.Parameters.Add(paramId);

                    var result = cmd.ExecuteNonQuery();
                    logger.Info($"updated {result} instances");
                }
            }
            catch(SQLiteException ex)
            {
                logger.Error(ex.Message);
                Console.WriteLine("Error DB " + ex.Message);
            }
        }
    }
}
